core/icons/mutator_icon.ts

```diff
@@ -0,0 +1,358 @@
+      this.sourceBlock.saveConnections(this.rootBlock);
+    };
+    saveConnectionsListener();
+    this.sourceBlock.workspace.addChangeListener(saveConnectionsListener);
+  }
+
+  /**
+   * Creates a change listener to add to the mini workspace which recomposes
+   * the block.
+   */
+  private createMiniWorkspaceChangeListener() {
+    return (e: Abstract) => {
+      if (!this.shouldIgnoreMutatorEvent(e) && !this.updateWorkspacePid) {
+        this.updateWorkspacePid = setTimeout(() => {
+          this.updateWorkspacePid = null;
+          this.recomposeSourceBlock();
+        }, 0);
+      }
+    };
+  }
+
+  /**
+   * Returns true if the given event is not one the mutator needs to
+   * care about.
+   */
+  private shouldIgnoreMutatorEvent(e: Abstract) {
+    return (
+      e.isUiEvent ||
+      e.type === eventUtils.CREATE ||
+      (e.type === eventUtils.CHANGE &&
+        (e as BlockChange).element === 'disabled')
+    );
+  }
+
+  /** Recomposes the source block based on changes to the mini workspace. */
+  private recomposeSourceBlock() {
+    if (!this.rootBlock) return;
+
+    const existingGroup = eventUtils.getGroup();
+    if (!existingGroup) eventUtils.setGroup(true);
+
+    const oldExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+    this.sourceBlock.compose!(this.rootBlock);
+    const newExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+
+    if (oldExtraState !== newExtraState) {
+      eventUtils.fire(
+        new (eventUtils.get(eventUtils.BLOCK_CHANGE))(
+          this.sourceBlock,
+          'mutation',
+          null,
+          oldExtraState,
+          newExtraState
+        )
+      );
+    }
+
+    eventUtils.setGroup(existingGroup);
+  }
+
+  /** @internal */
+  getWorkspace(): WorkspaceSvg | undefined {
+    return this.miniWorkspaceBubble?.getWorkspace();
+  }
+
+  /**
+   * Reconnects the given connection to the mutated input on the given block.
+   */
+  static reconnect(
+    connectionChild: Connection,
+    block: Block,
+    inputName: string
+  ): boolean {
+    if (!connectionChild || !connectionChild.getSourceBlock().workspace) {
+      return false; // No connection or block has been deleted.
+    }
+    const connectionParent = block.getInput(inputName)!.connection;
+    const currentParent = connectionChild.targetBlock();
+    if (
+      (!currentParent || currentParent === block) &&
+      connectionParent &&
+      connectionParent.targetConnection !== connectionChild
+    ) {
+      if (connectionParent.isConnected()) {
+        // There's already something connected here.  Get rid of it.
+        connectionParent.disconnect();
+      }
+      connectionParent.connect(connectionChild);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns the parent workspace of a workspace that is inside a mutator,
+   * taking into account wither it is a flyout.
+   */
+  static findParentWs(workspace: WorkspaceSvg): WorkspaceSvg | null {
```

Comment:
This seems to be called only from `blocks/procedures.js`; why not move it there—especially as it seems to have nothing to do with `MutatorIcon`?

Reply:
I will give this a look to see if this is procedure-specific, or if external developers could want to use it in a way that has nothing to do with procedures.

Reply:
I don't think this is procedure specific. Procedures use this to change something about the outer workspace in from a field validator on one of the mutator / mini workspace blocks, which could be done by non-procedure blocks.

Unless you can think of a third option for where to export it from, I think it makes more sense on `MutatorIcon` than it does in `procedures` (even though having it on `MutatorIcon` isn't great).